tests/cocotb/test_tcdm_subsys.py

```diff
@@ -10,29 +10,49 @@
 # ---------------------------------
 
 import cocotb
-from cocotb.triggers import RisingEdge, Timer
 from cocotb.clock import Clock
 from cocotb_test.simulator import run
-from decimal import Decimal
 import pytest
 import snax_util
+import math
 
 # Configurable design time parameters
 NARROW_DATA_WIDTH = 64
+WIDE_DATA_WIDTH = 512
 TCDM_DEPTH = 64
-NR_BANKS = 8
+NR_BANKS = 32
 NUM_INPUT = 2
 
+# Configuration checker for the DMA wide
+# TCDM interconnection because
+# We need to make sure that the memory
+# and interconnect matches the addressing
+TCDM_SIZE = int(NR_BANKS * TCDM_DEPTH * NARROW_DATA_WIDTH / 8)
+TCDM_ADDR_WIDTH = math.ceil(math.log2(TCDM_SIZE))
+TCDM_MEM_ADDR_WIDTH = math.ceil(math.log2(TCDM_DEPTH))
+NR_BANKS_PER_SUPERBANK = WIDE_DATA_WIDTH / NARROW_DATA_WIDTH
+NR_SUPER_BANKS = int(NR_BANKS / NR_BANKS_PER_SUPERBANK)
+WIDE_TCDM_PORTS = math.ceil(math.log2(NR_SUPER_BANKS))
+WIDE_BYTE_OFFSET = math.ceil(math.log2(int(WIDE_DATA_WIDTH / 8)))
+
+assert TCDM_ADDR_WIDTH >= (
+    TCDM_MEM_ADDR_WIDTH + WIDE_TCDM_PORTS + WIDE_BYTE_OFFSET
+), "Error config! Make sure to satisfy the assertion equation"
```

Comment:
Sorry, why this assertation?

Reply:
Let me try to explain haha but I figure it will take a whole document to describe it. However, if you go to this [line ](https://github.com/KULeuven-MICAS/snitch_cluster/blob/a76d915ac9977914b1bd613777c84475b6948f33/hw/snitch_cluster/src/snitch_tcdm_interconnect.sv#L92) in the code. It has this vector slicing mechanism:

```verilog
addr: req_i[i].q.addr[ByteOffset+SelWidth+:MemAddrWidth],
```

So the problem is the TCDM address bit width may be smaller than the DMA TCDM interconnect bit-width. Note that these are all calculated inside the TCDM (automatically) based on the original module.

The TCDM address (`TCDM_ADDR_WIDTH`) is the $\log_2 (\textrm{TCDM Size})$ where the TCDM size is the total number of bytes of the entire TCDM. The `TCDM_MEM_ADDR_WIDTH` is the address width of a single bank. `WIDE_TCDM_PORTS` is the number of bits to select one of the ports of the DMA's TCDM interconnect. Finally, the `WIDE_BYTE_OFFSET` is the byte offset in number of bits of the wide bit width.

So! Maybe not be clear for now but if you think about it, the address width of the TCDM port `TCDM_ADDR_WIDTH` needs to be greater than or equal to the sum of the address width of a single bank (`TCDM_MEM_ADDR_WIDTH`), the number of bits to select a TCDM port (`WIDE_TCDM_PORTS`), and the byte offset for the wide data width (`WIDE_BYTE_OFFSET`).